rotkehlchen/chain/ethereum/modules/optimism_gateway/decoder.py

```diff
@@ -0,0 +1,114 @@
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.fval import FVal
+from rotkehlchen.types import ChainID, ChecksumEvmAddress, EvmTokenKind, EvmTransaction
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+GATEWAY_ADDRESS = string_to_evm_address('0x99C9fc46f92E8a1c0deC1b1747d010903E884bE1')
+
+ERC20_DEPOSIT_INITIATED = b'q\x85\x94\x02z\xbdN\xae\xd5\x9f\x95\x16%c\xe0\xccm\x0e\x8d[\x86\xb1\xc7\xbe\x8b\x1b\n\xc34=\x03\x96'  # noqa: E501
+ETH_DEPOSIT_INITIATED = b'5\xd7\x9a\xb8\x1f+ \x17\xe1\x9a\xfb\\Uqw\x88wx-z\x87\x86\xf5\x90\x7f\x93\xb0\xf4p/O#'  # noqa: E501
+ERC20_WITHDRAWAL_FINALIZED = b'<\xee\xe0l\x1e7d\x8f\xcb\xb6\xedR\xe1{>\x1f\'Z\x1f\x8c{"\xa8K+\x84s$1\xe0F\xb3'  # noqa: E501
+ETH_WITHDRAWAL_FINALIZED = b'*\xc6\x9e\xe8\x04\xd9\xa7\xa0\x98BI\xf5\x08\xdf\xab|\xb2SKF[l\xe1X\x0f\x99\xa3\x8b\xa9\xc5\xe61'  # noqa: E501
+
+
+class BridgeData(NamedTuple):
+    amount: FVal
+    asset: CryptoAsset
+    source_address: ChecksumEvmAddress
+    destination_address: ChecksumEvmAddress
+
+
+def get_bridge_data_for_eth(tx_log: EvmTxReceiptLog) -> BridgeData:
+    eth_crypto_asset = A_ETH.resolve_to_crypto_asset()
+    raw_amount = hex_or_bytes_to_int(tx_log.data[:32])
+    return BridgeData(
+        amount=asset_normalized_value(raw_amount, eth_crypto_asset),
+        asset=eth_crypto_asset,
+        source_address=hex_or_bytes_to_address(tx_log.topics[1]),
+        destination_address=hex_or_bytes_to_address(tx_log.topics[2]),
+    )
+
+
+def get_bridge_data_for_erc20(tx_log: EvmTxReceiptLog) -> BridgeData:
+    ethereum_token_address = hex_or_bytes_to_address(tx_log.topics[1])
+    token = EvmToken(evm_address_to_identifier(
+        address=ethereum_token_address,
+        chain_id=ChainID.ETHEREUM,
+        token_type=EvmTokenKind.ERC20,
+    ))
+    raw_amount = hex_or_bytes_to_int(tx_log.data[32:64])
+    return BridgeData(
+        amount=asset_normalized_value(raw_amount, token),
+        asset=token,
+        source_address=hex_or_bytes_to_address(tx_log.topics[3]),
+        destination_address=hex_or_bytes_to_address(tx_log.data[:32]),
+    )
+
+
+class OptimismGatewayDecoder(DecoderInterface):
+    def _decode_bridge(  # pylint: disable=no-self-use
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: list[ActionItem],  # pylint: disable=unused-argument
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        if tx_log.topics[0] in {ETH_DEPOSIT_INITIATED, ETH_WITHDRAWAL_FINALIZED}:
+            bridge_data = get_bridge_data_for_eth(tx_log)
+        elif tx_log.topics[0] in {ERC20_DEPOSIT_INITIATED, ERC20_WITHDRAWAL_FINALIZED}:
+            bridge_data = get_bridge_data_for_erc20(tx_log)
```

Comment:
`get_bridge_data_for_erc20 ` and `get_bridge_data_for_eth` are similar functions with the only difference being a single line.

Perhaps combine into a single function and handle the condition in there